File: frontend/src/types/order.ts
export type OrderType = "dine_in" | "takeaway" | "delivery";
export type OrderStatus = "draft" | "completed" | "voided";
export type PaymentMethod = "cash" | "debit" | "credit" | "qris" | "other";
export type VoidReason = "Salah Input" | "Pelanggan Batal";

export interface OrderItem {
  id: string;
  order_id?: string;
  menu_item_id: string;
  quantity: number;
  price: number;
  notes?: string;
  is_void?: boolean;
  void_reason?: string;
  created_at?: string;
  menu_item?: { name: string; base_price: number };
}

export interface Order {
  id: string;
  shift_id?: string;
  user_id: string;
  type: OrderType;
  table_number?: string;
  tax_enabled: boolean;
  tax_rate: number;
  tax_amount: number;
  subtotal: number;
  grand_total: number;
  payment_method: PaymentMethod;
  amount_received?: number;
  change_amount?: number;
  status: OrderStatus;
  is_synced: boolean;
  original_timestamp?: string;
  created_at: string;
  updated_at: string;
  items?: OrderItem[];
}

export interface Payment {
  method: PaymentMethod;
  amount: number;
  change?: number;
}

File: backend/internal/models/order.go
package models

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type OrderType string
type PaymentMethod string
type OrderStatus string

const (
	OrderTypeDineIn   OrderType = "dine_in"
	OrderTypeTakeaway OrderType = "takeaway"
	OrderTypeDelivery OrderType = "delivery"

	PaymentCash   PaymentMethod = "cash"
	PaymentDebit  PaymentMethod = "debit"
	PaymentCredit PaymentMethod = "credit"
	PaymentQRIS   PaymentMethod = "qris"
	PaymentOther  PaymentMethod = "other"

	OrderDraft     OrderStatus = "draft"
	OrderCompleted OrderStatus = "completed"
	OrderVoided    OrderStatus = "voided"
)

type Order struct {
	ID                uuid.UUID     `gorm:"type:uuid;primaryKey" json:"id"`
	ShiftID           *uuid.UUID    `gorm:"type:uuid;index" json:"shift_id,omitempty"`
	UserID            uuid.UUID     `gorm:"type:uuid;not null;index" json:"user_id"`
	Type              OrderType     `gorm:"type:order_type;not null" json:"type"`
	TableNumber       *string       `gorm:"type:varchar(20)" json:"table_number,omitempty"`
	TaxEnabled        bool          `gorm:"not null;default:false" json:"tax_enabled"`
	TaxRate           float64       `gorm:"type:decimal(5,2);check:tax_rate >= 0" json:"tax_rate"`
	TaxAmount         float64       `gorm:"type:decimal(12,2);default:0" json:"tax_amount"`
	Subtotal          float64       `gorm:"type:decimal(12,2);not null;check:subtotal >= 0" json:"subtotal"`
	GrandTotal        float64       `gorm:"type:decimal(12,2);not null;check:grand_total >= 0" json:"grand_total"`
	PaymentMethod     PaymentMethod `gorm:"type:payment_method;not null" json:"payment_method"`
	AmountReceived    *float64      `gorm:"type:decimal(12,2);check:amount_received >= 0" json:"amount_received,omitempty"`
	ChangeAmount      *float64      `gorm:"type:decimal(12,2);check:change_amount >= 0" json:"change_amount,omitempty"`
	Status            OrderStatus   `gorm:"type:order_status;not null;default:'completed'" json:"status"`
	IsSynced          bool          `gorm:"not null;default:true" json:"is_synced"`
	OriginalTimestamp *time.Time    `json:"original_timestamp,omitempty"`
	CreatedAt         time.Time     `gorm:"not null;default:now()" json:"created_at"`
	UpdatedAt         time.Time     `gorm:"not null;default:now()" json:"updated_at"`

	User  *User       `gorm:"foreignKey:UserID" json:"user,omitempty"`
	Shift *Shift      `gorm:"foreignKey:ShiftID" json:"shift,omitempty"`
	Items []OrderItem `gorm:"foreignKey:OrderID;constraint:OnDelete:CASCADE" json:"items,omitempty"`
}

func (o *Order) BeforeCreate(tx *gorm.DB) error {
	if o.ID == uuid.Nil {
		o.ID = uuid.New()
	}
	return nil
}
